### PR TITLE
improve release page with extra info

### PR DESCRIPTION
**Pull request by lucabello**
## Summary

The current status page at https://releases.juju.is/ could be enriched with extra useful data. This PR aims to improve the page with CI and CharmHub information, to make it a more accurate representation of the team's repositories and a better self-monitoring tool.

Here you can see a preview of the changes being proposed, going from:

to:

(the images' purpose is to show how the information would be structured)

### What it would add

This PR introduces two main additions: CI status and CharmHub releases.

The current prototype presents the CI information in line with the rest; this doesn't thicken the row (because of the "Latest Releases") and allows teams to monitor their repositories more effectively.

The CharmHub releases would be displayed in a collapsible row, to avoid cluttering the screen, and show the current revision for each release track and channel, together with its release date. This information can be critical when multiple charms need to be promoted from `edge` to `stable`, as it allows monitoring the promotion status across all repositories (which is currently not possible anywhere else).  

### How it would happen

The releases page includes a wide variety of repositories: some are charms, some run CIs, some are more bare-bones. Moreover, different teams can use different nomenclatures for CI steps and release channels; it's not possible to know a priori which CI is relevant and which isn't.

The idea is to have the teams "opt-in" by adding the CI badges in each README.md; [releasegen](https://github.com/jnsgruk/releasegen) will then scrape the README.md of each project and visualize the status of those steps only.  
CharmHub releases would follow a similar approach; to show them in the page, the README.md should contain the CharmHub badge (found at `https://charmhub.io/<charm-name>/badge.svg`)
This way, the repository maintainers can have more control on this process, and we can avoid to "automatically detect" a charm through different repository structures and nomenclatures. If the badge is not found in the README.md, the button would be disabled.

### releasegen

It's clear that the data received by this repo will need to be adjusted for this use case; I'll open a PR for that soon and link it here.  

## Final Notes

Any comment or suggestion is very welcome!

### CODE REVIEW

File: layouts/index.html [hunk 1]
```diff
@@ -16,10 +16,21 @@ <h2>Charm Engineering Releases</h2>
...
-              Launchpad project groups. Github projects must have at least one Github Release,
-              Launchpad projects are scraped for tags. The latest three releases are shown for each
+              Launchpad project groups. Github projects must not be Archived, Launchpad projects
+              are scraped for tags. The latest three releases are shown for each
               repository.
             </p>
+            <p class="u-no-max-width">
+              For GitHub badges to be shown, add them in your README; similarly, to display the
+              extra CharmHub release information, add the CharmHub badge to your charm
+              repository.
```
Comment (jnsgruk) [inline layouts/index.html]: Might be worth including a link or something to let people know how to do this?

File: layouts/partials/css/icons.css [hunk 2]
```diff
@@ -0,0 +1,7 @@
+.inline-svg {
+  display: inline-block;
+  height: 1.15rem;
+  width: 1.15rem;
+  top: 0.15rem;
+  position: relative;
+}
```
Comment (jnsgruk) [inline layouts/partials/css/icons.css]: I think you can just include this in the existing `overrides.css` file

File: layouts/partials/head.html [hunk 3]
```diff
@@ -26,5 +26,6 @@
   <style>
     {{- partial "css/vanilla.css" . | safeCSS -}}
     {{- partial "css/overrides.css" . | safeCSS -}}
+    {{- partial "css/icons.css" . | safeCSS -}}
```
Comment (jnsgruk) [inline layouts/partials/head.html]: If you take the above advice, this can go :)

File: layouts/partials/js/expand.js [hunk 4]
```diff
@@ -0,0 +1,95 @@
+// Setup all expandable tables on the page.
+var tables = document.querySelectorAll('.p-table--expanding')
```
Comment (jnsgruk) [inline layouts/partials/js/expand.js]: Please use `const` or `let` throughout this file - ideally `const`.

File: layouts/partials/js/expand.js [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+    // Adjust the text of the toggle button
+    if (show) {
+      element.innerHTML = element.getAttribute('data-shown-text')
+    } else {
+      element.innerHTML = element.getAttribute('data-hidden-text')
+    }
+
+    target.setAttribute('aria-hidden', !show)
+  }
+}
```
Comment (jnsgruk) [inline layouts/partials/js/expand.js]: This function (and the others) should be written using the ES6 style declarations. See the `tabs.css` for examples.

```javascript
const toggleExpanded = (element, show ) => { ... }
```

File: layouts/partials/js/expand.js [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+    for (const b of expandButtons) {
+      if (!b.disabled) {
+        toggleExpanded(b, !isTargetActive)
+      }
+    }
+  },
+  false)
+
+// Add control to only show charms in the table
+var repositoryRows = document.querySelectorAll(".repository-row")
```
Comment (jnsgruk) [inline layouts/partials/js/expand.js]: Refactor this global into the `addEventListener` callback - it's only used once

File: layouts/partials/js/expand.js [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+var toggleCharmsOnly = document.getElementById('toggle-charms-only')
```
Comment (jnsgruk) [inline layouts/partials/js/expand.js]: Please change this to a `const`

```suggestion
const toggleCharmsOnly = document.getElementById('toggle-charms-only')
```

File: layouts/partials/js/expand.js [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+        if (!isCharm) {
+            r.style.display = r.style.display === 'none' ? '' : 'none'
+        }
+    }
+  },
+  false)
+
+
+
+
```
Comment (jnsgruk) [inline layouts/partials/js/expand.js]: You can use `prettier` to format this whole file and remove trailing whitespace, enforce style, etc.

Comment (jnsgruk) [reply layouts/partials/js/expand.js]: I should probably include a `prettierrc` in this repo, but fwiw these are my settings for `prettier` in VSCode:

```json
{
"prettier.printWidth": 99,
"prettier.semi": false,
"prettier.trailingComma": "es5",
}
  ```

File: layouts/partials/repo.html [hunk 9]
```diff
@@ -1,8 +1,17 @@
-<tr id="{{ .repo.name }}">
+<tr id="{{ .repo.name }}" class="repository-row">
   <td class="repository">
-    <a href="{{ .repo.url }}" class="repo-name">{{ .repo.name }}</a>
+    {{ .repo.name }}
+    <a href="{{ .repo.url }}" class="repo-icon icon-github">{{ partial "simpleicons.html" "github" }}</a>
```
Comment (jnsgruk) [inline layouts/partials/repo.html]: What about Launchpad repos?

File: layouts/partials/simpleicons.html [hunk 10]
```diff
@@ -0,0 +1,6 @@
+<span class="inline-svg" >
+{{- $fname:=print "simpleicons/" . ".svg" -}}
+{{- $path:="<path" -}}
+{{- $fill:="<path fill=\"currentColor\"" -}}
+{{ replace (readFile $fname) $path $fill | safeHTML }}
+</span>
```
Comment (jnsgruk) [inline layouts/partials/simpleicons.html]: This feels like a complicated way of including an svg on a page to me? Am I missing something?

**Review by jnsgruk — CHANGES_REQUESTED** [review]
Thanks for doing this Luca! We've got a few adjustments to make in the code, and I have some observations on the overall layout:

image

In the above image:

1) The alignment seems off in the table columns to me?
2) We should probably do a `s/Continuous Integration/CI/` in that header for brevity
3) I think I'd keep the charm name as a link to the Github repo, and drop the Github logo, that will also mean moving the charmhub logo into the expanded section
4) The 'Show' button is quite large - perhaps replace with a simple `...` or down arrow?

Next..

image

The `Charms Only` and `Expand All` buttons work really great, so nice work on that! I do think we need to find a neater place for them, though. They look a bit clunky where they are now :)

Let me know what you think!

**Comment by lucabello** [comment]
Thanks for this first review!
The layout already looks better now, as I:
* moved the icons in a separate column next to the repo name
* resize columns for a better layout
So make sure to pull the newest changes :)

The header of the CI column is "Continuous Integration" just because the column needs to be large enough for badges; it looks uglier on teams where that column is fully empty (it shouldn't be, as people update their READMEs), but changing it makes it uglier on columns where the badge is there. I don't really have a problem changing it though if you still think it looks better!

I moved the links to a separate small column next to the name; I'd be okay to remove it altogether, but I'd like to have a CharmHub link somewhere regardless. Is there a better place for it?

(as per the JavaScript changes, I'm on to them!)

**Comment by jnsgruk** [comment]
> The header of the CI column is "Continuous Integration" just because the column needs to be large enough for badges; it looks uglier on teams where that column is fully empty (it shouldn't be, as people update their READMEs), but changing it makes it uglier on columns where the badge is there. I don't really have a problem changing it though if you still think it looks better!

Ah fair enough - maybe just "Status"? I think the badges should be centre aligned too, given that they're all such different widths :)

It might be worth adding some placeholder in that column (perhaps in greyed out text) where there are no badges present too?

I'll take a look at the style changes shortly and come back!

**Comment by jnsgruk** [comment]
So that *is* looking better - thanks for the quick update!

You can see the effects of that wide CI column best here:

image

To me the alignment of the Latest Releases column still looks off compared to https://releases.juju.is, though?

And we still need to find somewhere less glaring for those expander/show/hide buttons to live - I'll have a think, but maybe glance through the examples on the vanilla page -- maybe a dropdown in the nav in the very top right under a :gear: icon or something?

**Comment by lucabello** [comment]
I've made some further changes and things should be more in line with your comments now :)

Screenshot_20230213_170511

The repository link is in the name, and I moved the CharmHub link in the collapsed page, as it's only relevant for charms anyway. Should be much better now! 

Also the CI badges are now centered and there is some placeholder!

For the buttons outside the table they are indeed awkward there; I'll try to put that ⚙️ to show them somewhere above the table possibly!

File: layouts/partials/js/expand.js (hunk 4 above)
Comment (lucabello) [reply layouts/partials/js/expand.js]: Yeah, this was taken from vanillaframeworks examples, but I fixed it now!

**Comment by lucabello** [comment]
There is now a dropdown for those two "table filters" at the top-right corner of the table!

JavaScript files have been improved and fixed with `prettier`, so all major concerns should be addressed now :)

File: layouts/partials/js/dropdown.js [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+    }
+  }
+}
+
+/**
+  Attaches event listeners for the menu toggle open and close click events.
+  @param {HTMLElement} menu The menu container element.
+*/
+
+function setupContextualMenu(menu) {
```
Comment (jnsgruk) [inline layouts/partials/js/dropdown.js]: Can you convert the functions in this file to ES6 style "fat-arrow" functions like the other file please? :)

Comment (lucabello) [reply layouts/partials/js/dropdown.js]: I knew I forgot some :) Done!

**Comment by jnsgruk** [comment]
This is starting to look really smart, thanks Luca! 

A couple more points:

image

Can you remove the border box around that paragraph and just merge it into the paragraph above? No need for it to be some prominent. Perhaps we could put the whole section about how to be included into an [Accordion](https://vanillaframework.io/docs/patterns/accordion) element titled "Need help getting your repo included here?" that is collapsed by default?

Another uber-nit...

image

I think the columns should be in the order:

Name / Latest Release / Status / Commits Since / [info symbol] (see [`p-icon-information`](https://vanillaframework.io/docs/patterns/icons))

Note that I dropped the "Charmhub" column and replaced with the "info" icon -- I think *if* there are badges (i.e. not the placeholder), then we should render the expand button, and if not the entire expanded table row and button should not be rendered

Finally... can we centre the "Commits Since" and status columns :)

**Comment by jnsgruk** [comment]
Oh! And you broke my "eye" icon links ;-)

image

File: layouts/partials/repo.html [hunk 12]
```diff
@@ -1,6 +1,15 @@
...
-    <a href="{{ .repo.url }}" class="repo-name">{{ .repo.name }}</a>
+    <a href="{{ .repo.url }}">{{ .repo.name }}</a>
+  </td>
+  <td class="u-align--center cistages">
+    {{- if not (not .repo.ci) -}}
+      {{- range .repo.ci -}}
+        <a href="{{- . -}}"><img src="{{- . -}}/badge.svg"></a>
+      {{- end -}}
+    {{- else -}}
+    <img src="https://img.shields.io/badge/CI%20Placeholder-some%20status-lightgrey">
```
Comment (jnsgruk) [inline layouts/partials/repo.html]:
```suggestion
    <img src="https://img.shields.io/badge/CI%20Status-Unknown-lightgrey">
```

**Comment by lucabello** [comment]
Your comments have been addressed!

The expand button is tied to the repo holding a charm, not to the CI stages; I assumed you meant that for both CI and Expansion, you want the whole column to be hidden if there is no actionable information for all charms in a specific team; that's done now!

Basically the page will look the same as before for teams that don't modify their READMEs at all; I was in any case preparing a Discourse post to spread the news to people to publish after all the changes are merged.

A small note you might not have noticed, which maybe concerns more the `releasegen` PR; currently: 
1. repos that have no releases are filtered out and not displayed
2. archived repos are displayed

I changed this in the PR so that:
1. repos are displayed regardless of releases being there (some repos might have no GitHub release but hold important code (e.g. CI workflows) for which you will want to see the CI status). We could add an extra item to the dropdown menu to hide the repos without release if you deem it useful, but I think the default should be to show them
2. archived repos are not displayed; this is because an archived repo with failing CI / unreleased commits will show as "failing" forever, which is counterproductive when you want to look at the status of your repos; also here, I wouldn't mind adding a button to show archived repos as well, but I believe they should be hidden by default

Let me know what you think about these points, or if you have any other remarks on the page in general!

**Comment by lucabello** [comment]
As discussed outside this PR, repos without GitHub releases now display the latest commit in a similar fashion. Repos without any commits (or archived repos) will not be shown (`releasegen` will cut them off).

Things currently look like this:

(updated to reflect latest commits)

Screenshot_20230214_175619

File: .gitignore [hunk 13]
```diff
@@ -5,4 +5,10 @@
...
 # Temporary lock file while building
-/.hugo_build.lock
\ No newline at end of file
+/.hugo_build.lock
+
+# Node
+node_modules/
+package-lock.json
+package.json
+
```
Comment (jnsgruk) [inline .gitignore]: Why are these here? :)

File: layouts/index.html [hunk 14]
```diff
@@ -16,13 +33,27 @@ <h2>Charm Engineering Releases</h2>
...
+            <aside class="p-accordion">
+              <ul class="p-accordion__list">
+                <li class="p-accordion__group">
+                  <div role="heading" aria-level="3" class="p-accordion__heading">
+                    <button type="button" class="p-accordion__tab" id="accordion-help-tab" aria-controls="accordion-help" aria-expanded="false">Need help getting your repository fully included here?</button>
+                  </div>
+                  <section class="p-accordion__panel" id="accordion-help" aria-hidden="true" aria-labelledby="accordion-help-tab">
+                    <p>For GitHub badges to be shown for your repository, add them in the README; similarly,
+                    to display the extra CharmHub release information, add the CharmHub badge
+                    ( <code>![](https://charmhub.io/*charm-name*/badge.svg)</code> ) to your charm repository.</p>
```
Comment (jnsgruk) [inline layouts/index.html]:
```suggestion
                    <p p class="u-no-max-width">For GitHub badges to be shown for your repository, add them in the README; similarly,
                    to display the extra CharmHub release information in the expanding table rows, add the CharmHub badge
                    ( <code>![](https://charmhub.io/*charm-name*/badge.svg)</code> ) to your charm's README.</p>
```

**Review by jnsgruk — COMMENTED** [review]
This is looking really tidy now, thanks Luca! Just a couple of nits

File: layouts/index.html [hunk 15]
```diff
@@ -16,13 +33,27 @@ <h2>Charm Engineering Releases</h2>
             </p>
             <p class="u-no-max-width">
               For a release to show here, it must be part of one of the specified Github Teams or
-              Launchpad project groups. Github projects must have at least one Github Release,
-              Launchpad projects are scraped for tags. The latest three releases are shown for each
+              Launchpad project groups. Github projects must not be Archived, Launchpad projects
+              are scraped for tags. The latest three releases are shown for each
               repository.
             </p>
```
Comment (jnsgruk) [inline layouts/index.html]: I wonder if this para should also move into the accordion while we're tidying this up?

File: layouts/index.html [hunk 16]
```diff
@@ -16,13 +33,27 @@ <h2>Charm Engineering Releases</h2>
...
+            <aside class="p-accordion">
+              <ul class="p-accordion__list">
+                <li class="p-accordion__group">
+                  <div role="heading" aria-level="3" class="p-accordion__heading">
+                    <button type="button" class="p-accordion__tab" id="accordion-help-tab" aria-controls="accordion-help" aria-expanded="false">Need help getting your repository fully included here?</button>
```
Comment (jnsgruk) [inline layouts/index.html]:
```suggestion
                    <button type="button" class="p-accordion__tab" id="accordion-help-tab" aria-controls="accordion-help" aria-expanded="false">Need help getting your repository included here?</button>
```

File: .gitignore (hunk 13 above)
Comment (lucabello) [reply .gitignore]: Whops, I thought this was put there by prettier CLI, but it wasn't :)

File: layouts/index.html (hunk 15 above)
Comment (lucabello) [reply layouts/index.html]: Good point, I would agree it belongs there!

**Comment by lucabello** [comment]
Everything should be good now :) I would still wait for https://github.com/jnsgruk/releasegen/pull/1 to be merged before merging this one, just in case we end up restructuring the JSON

Once you confirm everything looks good, I'll also remove the `data.json` from the PR !

**Review by jnsgruk — APPROVED** [review]
Nice work! Let's hold on merging this until the releasegen PR is across the line, but thanks for your efforts here! Looks really smart!
